pkg/cache/v3/simple.go

```diff
@@ -402,54 +402,58 @@ func (cache *snapshotCache) CreateWatch(request *Request, sub Subscription, valu
 	info.lastWatchRequestTime = time.Now()
 	info.mu.Unlock()
 
-	var version string
+	createWatch := func(watch ResponseWatch) func() {
+		watchID := cache.nextWatchID()
+		cache.log.Debugf("open watch %d for %s %v from nodeID %q, version %q", watchID, request.GetTypeUrl(), sub.SubscribedResources(), nodeID, request.GetVersionInfo())
+		info.mu.Lock()
+		info.watches[watchID] = watch
+		info.mu.Unlock()
+		return cache.cancelWatch(nodeID, watchID)
+	}
+
+	watch := ResponseWatch{Request: request, Response: value, subscription: sub}
+
 	snapshot, exists := cache.snapshots[nodeID]
-	if exists {
-		version = snapshot.GetVersion(request.GetTypeUrl())
+	if !exists {
+		return createWatch(watch), nil
 	}
 
-	if exists {
+	version := snapshot.GetVersion(request.GetTypeUrl())
+	resources := snapshot.GetResourcesAndTTL(request.GetTypeUrl())
+
+	if request.GetVersionInfo() == version {
+		// Retrieve whether there are resources in the cache requested and currently unknown to the client.
 		knownResourceNames := sub.ReturnedResources()
-		diff := []string{}
-		for _, r := range request.GetResourceNames() {
-			if _, ok := knownResourceNames[r]; !ok {
-				diff = append(diff, r)
+		diff := false
+		if !sub.IsWildcard() {
+			// Check if a resource requested is currently not returned,
+			// for instance if it is newly subscribed
+			for r := range sub.SubscribedResources() {
+				if _, ok := knownResourceNames[r]; !ok {
+					diff = true
+					break
+				}
 			}
-		}
-
-		cache.log.Debugf("nodeID %q requested %s%v and known %v. Diff %v", nodeID,
-			request.GetTypeUrl(), request.GetResourceNames(), knownResourceNames, diff)
-
-		if len(diff) > 0 {
-			resources := snapshot.GetResourcesAndTTL(request.GetTypeUrl())
-			for _, name := range diff {
-				if _, exists := resources[name]; exists {
-					if err := cache.respond(context.Background(), request, value, resources, version, false); err != nil {
-						cache.log.Errorf("failed to send a response for %s%v to nodeID %q: %s", request.GetTypeUrl(),
-							request.GetResourceNames(), nodeID, err)
-						return nil, fmt.Errorf("failed to send the response: %w", err)
-					}
-					return func() {}, nil
+		} else {
+			// Check if a resource present in the snapshot is currently not returned,
```

Comment:
I think this is a good precautionary check but I'm curious whether we expect to ever find a diff here. Adding a resource to the snapshot should increment the version, shouldn't it?